### PR TITLE
fix(NODE-5201): prevent warning when default value for deprecation option is used

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

- Default options are no longer collected as a part of the `allOptions` map.  Since `allOptions` no longer contains defaults, it is now named `providedOptions`
- deprecation warnings are emitted inside `parseOptions`, not `setOption`, and are only emitted for non-default values

##### Is there new documentation needed for these changes?

No.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/connection_string.ts [hunk 1]
```diff
@@ -334,30 +330,29 @@ export function parseOptions(
     if (urlValue != null) {
       values.push(...urlValue);
     }
-    const defaultOptionsValue = DEFAULT_OPTIONS.get(key);
-    if (defaultOptionsValue != null) {
-      values.push(defaultOptionsValue);
-    }
-    allOptions.set(key, values);
+    allProvidedOptions.set(key, values);
```
Comment (nbbeeken) [inline src/connection_string.ts:333]: Consider tracking source, feel free to resolve

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/unit/mongo_client.test.js [hunk 2]
```diff
@@ -488,22 +488,16 @@ describe('MongoOptions', function () {
...
       });
       const clientViaUri = new MongoClient('mongodb://localhost?compressors=zlib,snappy');
-      expect(clientViaOpt.options)
-        .to.have.property('compressors')
-        .deep.equal(['zlib', 'snappy', 'none']);
-      expect(clientViaUri.options)
-        .to.have.property('compressors')
-        .deep.equal(['zlib', 'snappy', 'none']);
+      expect(clientViaOpt.options).to.have.property('compressors').deep.equal(['zlib', 'snappy']);
+      expect(clientViaUri.options).to.have.property('compressors').deep.equal(['zlib', 'snappy']);
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply test/unit/mongo_client.test.js:500]: Can you elaborate on this comment?  I'm not sure I understand the concern here - it seems like it's working as designed? 

- Options parsing constructs the options object that our driver uses.  If we fix a bug in options parsing related to compressors, and that will also affect the handshake but that's expected.
- We have tests to ensure that we don't break options parsing

Is there something different you're concerned with?

[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/mongo_client.test.js:500]: We are fixing the fact that 'none' makes its way down to the handshake construction and gets sent to the server. I feel like we could add a catch to make sure that 'none' doesn't get reintroduced to the handshake. If its reasonable, which I think with the recent handshake tests that came with faas we have a framework to work with

[the remark this answers is not in the document]
Comment (baileympearson) [reply test/unit/mongo_client.test.js:500]: are you looking for a test like:

```typescript
context('when compressors are set on the client options, () => {
  it('does not include `none` as an option on the handshake', () => .. );
});
```

[the remark this answers is not in the document]
Comment (baileympearson) [reply test/unit/mongo_client.test.js:500]: How is that different than the existing tests?

[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/mongo_client.test.js:500]: We don't have coverage for the handshake just options parsing, demonstrated by the fact that this change didn't fail anything

**Review by nbbeeken — APPROVED** [review]

**Review by durran — APPROVED** [review]

**Pull request merged**